pkg/perf/pubip_task.go

```diff
@@ -0,0 +1,238 @@
+			}
+		}
+		err = netlink.LinkSetDown(mv)
+		if err != nil {
+			return fmt.Errorf("failed to set link down: %w", err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to run public IP validation: %w", err)
+	}
+	return unusedIPs, nil
+}
+
+func isLeastValidNode(ctx context.Context, farmID uint32, sub *substrate.Substrate) (bool, error) {
+	nodes, err := sub.GetNodes(uint32(farmID))
+	if err != nil {
+		return false, fmt.Errorf("failed to get farm %d nodes: %w", farmID, err)
+	}
+	cl := GetZbusClient(ctx)
+	registrar := stubs.NewRegistrarStub(cl)
+	var nodeID uint32
+	for {
+		nodeID, err = registrar.NodeID(ctx)
+		if err == nil {
+			break
+		}
+		log.Err(err).Msg("failed to get node id")
+		time.Sleep(10 * time.Second)
+	}
+	for _, node := range nodes {
+		if node >= uint32(nodeID) {
+			continue
+		}
+		state, err := sub.GetPowerTarget(node)
+		if err != nil {
+			return false, fmt.Errorf("failed to get node %d power target: %w", node, err)
+		}
+		if state.Target.IsDown {
+			continue
+		}
+		n, err := sub.GetNode(node)
+		if err != nil {
+			return false, fmt.Errorf("failed to get node %d: %w", node, err)
+		}
+		ip, err := getValidNodeIP(n)
+		if err != nil {
+			return false, err
+		}
+		// stop at three and quiet output
+		err = exec.CommandContext(ctx, "ping", "-c", "3", "-q", ip).Run()
+		if err != nil {
+			log.Err(err).Msgf("failed to ping node %d", node)
+			continue
+		}
+		return false, nil
+	}
+	return true, nil
+}
+
+func getValidNodeIP(node *substrate.Node) (string, error) {
+	for _, inf := range node.Interfaces {
+		if inf.Name != "zos" {
+			continue
+		}
+		if len(inf.IPs) == 0 {
+			return "", fmt.Errorf("no private IP available on node %d", node.ID)
+		}
+		return inf.IPs[0], nil
+	}
+	return "", fmt.Errorf("failed to get private IP for node %d", node.ID)
+}
+
+func getIPWithRoute(publicIP substrate.PublicIP) (net.IP, []*net.IPNet, []*netlink.Route, error) {
+	ip, ipNet, err := net.ParseCIDR(publicIP.IP)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to parse IP %s: %w", publicIP.IP, err)
+	}
+	gateway := net.ParseIP(publicIP.Gateway)
+	if gateway == nil {
+		return nil, nil, nil, fmt.Errorf("failed to parse gateway %s: %w", publicIP.Gateway, err)
+	}
+	route, err := netlink.RouteGet(gateway)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get route to gateway %s", publicIP.Gateway)
+	}
+	routes := make([]*netlink.Route, 0)
+	for _, r := range route {
+		routes = append(routes, &r)
+	}
+	return ip, []*net.IPNet{ipNet}, routes, nil
+}
+
+func getRealPublicIP() (string, error) {
+	// for testing now, should change to cloudflare
+	req, err := http.Get("https://api.ipify.org/")
+	if err != nil {
+		return "", err
+	}
+	defer req.Body.Close()
+
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+func deleteIPAndRoutes(publicIP substrate.PublicIP, routes []*netlink.Route, macvlan netlink.Link) error {
```

Comment:
I really think you should just delee all Ips and routes blindly, so u start by listing the ips on this link and ALL routes inside the namespace and delete them.

This function should also be called before the test to make sure there are no left over ips on the device because of a crash